### PR TITLE
Add rope pos emb in fp32

### PR DESCRIPTION
**Pull request by hsiehjackson**
# What does this PR do ?

Support pure fp32 for rope pos embedding

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: examples/nlp/language_modeling/megatron_gpt_eval.py [hunk 1]
```diff
@@ -304,7 +305,7 @@ def main(cfg) -> None:
     # print("***************************")
 
     # Second method of running text generation, call trainer.predict [recommended]
-    bs = cfg.inference.batch_size #8 if fp8_enabled else 2
+    bs = cfg.inference.batch_size  # 8 if fp8_enabled else 2
```
Comment (yzhang123) [inline examples/nlp/language_modeling/megatron_gpt_eval.py:308]: do you need the comment # 8 if fp8_enabled else 2

File: examples/nlp/language_modeling/megatron_gpt_eval_zero_scrolls.py [hunk 2]
```diff
@@ -163,6 +163,9 @@ def main(cfg) -> None:
             if cfg.inference.get("apply_query_key_layer_scaling", None) is not None:
                 pretrained_cfg.apply_query_key_layer_scaling = cfg.inference.apply_query_key_layer_scaling
 
+            if cfg.get('enforce_fp32_pos_idx', None) is not None:
```
Comment (yzhang123) [inline examples/nlp/language_modeling/megatron_gpt_eval_zero_scrolls.py:166]: shall we have this as default? same as llama?

File: nemo/collections/nlp/models/language_modeling/megatron/gpt_model.py [hunk 3]
```diff
@@ -167,6 +167,7 @@ def __init__(
         ub_tp_comm_overlap=False,
         use_flash_attention=False,
         seq_len_interpolation_factor=None,
+        enforce_fp32_pos_idx=False,
```
Comment (yzhang123) [inline nemo/collections/nlp/models/language_modeling/megatron/gpt_model.py:170]: see above

File: nemo/collections/nlp/modules/common/megatron/position_embedding/alibi_relative_position_embedding.py [hunk 4]
```diff
@@ -49,7 +49,7 @@ def build_slopes(num_attention_heads, num_attention_heads_alibi):
     )
 
     # if torch.cuda.is_available():
-        # slopes = slopes.to(torch.cuda.current_device())
+    # slopes = slopes.to(torch.cuda.current_device())
```
Comment (yzhang123) [inline nemo/collections/nlp/modules/common/megatron/position_embedding/alibi_relative_position_embedding.py:52]: remove

File: nemo/collections/nlp/modules/common/megatron/position_embedding/alibi_relative_position_embedding.py [hunk 5]
```diff
@@ -66,7 +66,7 @@ def build_relative_position(max_seq_len, full=True):
         relative_position = torch.abs(memory_position - relative_position)  # (max_seq_len, max_seq_len)
 
     # if torch.cuda.is_available():
-        # relative_position = relative_position.to(torch.cuda.current_device())
+    # relative_position = relative_position.to(torch.cuda.current_device())
```
Comment (yzhang123) [inline nemo/collections/nlp/modules/common/megatron/position_embedding/alibi_relative_position_embedding.py:69]: remove

File: nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py [hunk 6]
```diff
@@ -25,7 +25,13 @@ class RotaryEmbedding(nn.Module):
     Implements Rotary Position Embedding from https://arxiv.org/abs/2104.09864.
     """
 
-    def __init__(self, dim: int, seq_len_interpolation_factor: int = None, pretrained_max_position_embeddings: int = None):
+    def __init__(
+        self,
+        dim: int,
+        seq_len_interpolation_factor: int = None,
+        pretrained_max_position_embeddings: int = None,
+        enforce_fp32_pos_idx: bool = False,
```
Comment (yzhang123) [inline nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py:33]: see above

**Review by yzhang123 — CHANGES_REQUESTED** [review]

**Review by ekmb — APPROVED** [review]

**Review by shan18 — APPROVED** [review]
lgtm

**Pull request merged**